src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java

```diff
@@ -0,0 +1,223 @@
+package com.uber.jenkins.phabricator.coverage;
+
+import com.google.common.annotations.VisibleForTesting;
+import hudson.FilePath;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.Project;
+import hudson.model.Run;
+import hudson.plugins.jacoco.ExecutionFileLoader;
+import hudson.plugins.jacoco.JacocoBuildAction;
+import hudson.plugins.jacoco.JacocoPublisher;
+import hudson.plugins.jacoco.report.CoverageReport;
+import hudson.remoting.VirtualChannel;
+import jenkins.MasterToSlaveFileCallable;
+import org.apache.tools.ant.DirectoryScanner;
+import org.jacoco.core.analysis.ICounter;
+import org.jacoco.core.analysis.ILine;
+import org.jacoco.core.analysis.IPackageCoverage;
+import org.jacoco.core.analysis.ISourceFileCoverage;
+import org.jacoco.core.analysis.ISourceNode;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides Jacoco coverage data
+ */
+@SuppressWarnings("unused")
+public class JacocoCoverageProvider extends CoverageProvider {
+
+    private static final int PERCENTAGE_UNAVAILABLE = -1;
+
+    @Override
+    public boolean hasCoverage() {
+        CoverageReport result = getCoverageResult();
+        return result != null && result.hasLineCoverage();
+    }
+
+    private CoverageReport getCoverageResult() {
+        JacocoBuildAction jacocoAction = getJacocoBuildAction();
+        if (jacocoAction == null) {
+            return null;
+        }
+        return jacocoAction.getResult();
+    }
+
+    private JacocoBuildAction getJacocoBuildAction() {
+        Run<?, ?> build = getBuild();
+        if (build == null) {
+            return null;
+        }
+
+        return build.getAction(JacocoBuildAction.class);
+    }
+
+    @Override
+    protected CodeCoverageMetrics getCoverageMetrics() {
+        return convertJacoco(getCoverageResult());
+    }
+
+    @Override
+    public Map<String, List<Integer>> readLineCoverage() {
+        JacocoBuildAction jacocoAction = getJacocoBuildAction();
+        if (jacocoAction == null) {
+            return null;
+        }
+
+        PathResolver pathResolver = new PathResolver(getWorkspace(), getSourceDirs());
+
+        HashMap<String, List<Integer>> lineCoverage = new HashMap<String, List<Integer>>();
+
+        try {
+            ExecutionFileLoader executionFileLoader = jacocoAction.getJacocoReport().parse(null, null);
```

Comment:
If we only need the `getJacocoReport()`, we should not be gating this feature on the presence of a jacoco build step. All we need is for the plugin to be available in jenkins so we have the runtime components to parse a coverage result. We should be able to use it to parse the result and provide coverage. The cobertura provider does not need the coberyura plugin to be applied to the job either, so we should mirror the same behavior by adding a config option to the phab jenkins plugint o accept the path to the jacoco xml report.

Reply:
No worries. Feel free to leave it this way. We will add a way to not have the jacoco plugin added to the job and still use the plugin in a separate change.